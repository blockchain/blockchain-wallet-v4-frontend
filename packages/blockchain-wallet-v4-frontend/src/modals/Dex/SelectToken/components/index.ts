export * from './Header'
export * from './SearchField'
export * from './TokenList'
