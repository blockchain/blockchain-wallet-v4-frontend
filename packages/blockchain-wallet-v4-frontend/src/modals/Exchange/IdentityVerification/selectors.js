import { path } from 'ramda'
import { selectors } from 'data'

export const getData = state => ({
  helperDomain: path(
    ['walletOptionsPath', 'data', 'domains', 'walletHelper'],
    state
  ),
  step: selectors.components.identityVerification.getVerificationStep(state),
  steps: selectors.components.identityVerification.getSteps(state)
})
