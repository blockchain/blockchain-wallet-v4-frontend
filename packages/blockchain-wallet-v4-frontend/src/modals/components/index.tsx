import React, { useState } from 'react'
import { FormattedMessage } from 'react-intl'
import styled, { css } from 'styled-components'

import { Button, Icon, Image, SpinningLoader, Text } from 'blockchain-info-components'
import { FlyoutWrapper } from 'components/Flyout'
import { OBInstitution } from 'data/types'

const BankWrapper = styled(FlyoutWrapper)`
  padding: 37px 0 34px;
  height: 100%;
  width: 100%;
  display: flex;
  flex-direction: column;
`
const NavText = styled(Text)`
  display: flex;
  align-items: center;
  margin-bottom: 24px;
  padding-left: 40px;
`

const Section = styled.div`
  display: flex;
  flex-direction: column;
  flex: 2;
  align-items: center;
  padding: 20px;
`
const QrContainer = styled.div`
  width: 150px;
  height: 150px;
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 40px auto 0;
  padding: 15px;
  border: 2px solid ${(p) => p.theme.blue600};
  border-radius: 4px;

  & img {
    width: 150px;
  }
`
const WaitingContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  opacity: 0;
  transition: opacity 0.5s ease-out;
  margin-bottom: 15px;

  & > div:first-child {
    margin-right: 10px;
  }

  &.active {
    opacity: 1;
  }
`
const LinkOptionsWrapper = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
`
const LogoImage = styled.img`
  margin-bottom: 32px;
  max-height: 60px;
`

interface ScanWithPhoneType {
  children: React.ReactChild
  readonly logo?: string
  readonly qrCode?: string
}
const ScanWithPhone = ({ children, logo, qrCode }: ScanWithPhoneType) => {
  return (
    <Section>
      {logo && <LogoImage src={logo} />}
      <Text weight={600} size='20px' color='grey900'>
        {children}
      </Text>
      <Text weight={500} size='14px' color='grey600'>
        <FormattedMessage
          id='modals.brokerage.use_phone_camera'
          defaultMessage='Use your phone’s camera to scan the QR code.'
        />
      </Text>
      <QrContainer>
        {qrCode ? (
          <img alt='Use your phone’s camera to scan the QR code.' src={qrCode} />
        ) : (
          <SpinningLoader width='30px' height='30px' />
        )}
      </QrContainer>
    </Section>
  )
}

const BankWaitIndicator = ({ qrCode }: { readonly qrCode?: string }) => {
  const [waitCount, setWaitCount] = useState(0)

  if (qrCode) {
    setTimeout(() => {
      setWaitCount(waitCount + 1)
    }, 3e4)
  }

  return (
    <WaitingContainer className={qrCode ? 'active' : ''}>
      <SpinningLoader width='10px' height='10px' borderWidth='3px' />
      <Text size='14px' weight={500}>
        {waitCount === 0 && (
          <FormattedMessage
            id='modals.brokerage.waiting_to_hear'
            defaultMessage='Waiting to hear from your bank'
          />
        )}
        {waitCount > 0 && (
          <FormattedMessage
            id='modals.brokerage.this_can_take_a_while'
            defaultMessage='This can take several minutes, hold tight!'
          />
        )}
      </Text>
    </WaitingContainer>
  )
}

const StyledButton = styled(Button)`
  margin: 20px 0 0;
  display: unset;
`
const LinkViaDesktop = ({
  authUrl,
  children,
  onClick = () => {
    // do nothing.
  }
}: {
  authUrl?: string
  children: React.ReactChild
  onClick?: () => void
}) => {
  if (!authUrl) return null
  return (
    <Section>
      <Text weight={600} size='20px' color='grey900'>
        {children}
      </Text>
      <StyledButton
        data-e2e='yapilyBankLink'
        nature='empty-blue'
        onClick={() => {
          window.open(authUrl, '_blank')
          onClick() // additional callback from implementing component
        }}
      >
        <FormattedMessage
          id='modals.brokerage.continue_in_browser'
          defaultMessage='Continue in browser'
        />
      </StyledButton>
    </Section>
  )
}

const StyledText = styled(Text)`
  width: 300px;
`

const Wrapper = styled.div`
  width: 100%;
  height: 100%;
  align-items: center;
  justify-content: center;
  display: flex;
  flex-direction: column;
  text-align: center;
`

const BROKERAGE_INELIGIBLE = 'BROKERAGE_INELIGIBLE'

const IneligibleErrorMessage = () => (
  <StyledText size='16px' weight={400}>
    <FormattedMessage
      id='modals.brokerage.ineligible_error'
      defaultMessage='You are not eligible to make deposits and withdrawals with this currency.'
    />
  </StyledText>
)

// getting ready, processing, loading
export enum LoadingTextEnum {
  GETTING_READY = 'Getting Ready...',
  LOADING = 'Loading...',
  PROCESSING = 'Processing...'
}
interface Props {
  text: LoadingTextEnum
}
const Loading = ({ text }: Props) => {
  return (
    <Wrapper>
      <SpinningLoader />
      <Text weight={600} color='grey600' style={{ marginTop: '24px' }}>
        {text === LoadingTextEnum.LOADING && (
          <FormattedMessage id='copy.loading' defaultMessage='Loading...' />
        )}
        {text === LoadingTextEnum.GETTING_READY && (
          <FormattedMessage id='loader.message.gettingready' defaultMessage='Getting Ready...' />
        )}
        {text === LoadingTextEnum.PROCESSING && (
          <FormattedMessage id='modals.simplebuy.processing' defaultMessage='Processing…' />
        )}
      </Text>
    </Wrapper>
  )
}

const SpinnerContainer = styled.div`
  transform: translate(47px, 35px);
  border: 5px solid white;
  border-radius: 50%;
  background-color: white;
`
const HeadingText = styled(Text)`
  font-weight: 600;
  font-size: 20px;
  margin-top: 22px;
`
const BodyText = styled(Text)`
  font-size: 14px;
  font-weight: 500;
  margin: 5px 30px;
`
const LoadingUpdating = () => {
  return (
    <Wrapper>
      <SpinnerContainer>
        <SpinningLoader borderWidth='7px' height='32px' width='32px' />
      </SpinnerContainer>
      <Image name='blockchain-logo-circle' width='106px' />
      <HeadingText color='grey900'>
        <FormattedMessage
          defaultMessage='Updating Your Wallet...'
          id='modals.brokerage.updating_your_wallet'
        />
      </HeadingText>
      <BodyText color='grey600'>
        <FormattedMessage
          defaultMessage='This could take up to 30 seconds. Please do not go back or close the app.'
          id='modals.brokerage.this_could_take'
        />
      </BodyText>
    </Wrapper>
  )
}

interface BankIconProps {
  url: string
}

const BankRow = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  border: 0px solid ${(p) => p.theme.grey000};
  border-bottom-width: 1px;
  padding: 28px 40px;

  & > div {
    display: flex;
    flex-direction: row;
    justify-content: flex-start;
    align-items: center;
  }

  ${(props) =>
    props.onClick &&
    css`
      cursor: pointer;
      * {
        cursor: pointer;
      }
      &:hover {
        background-color: ${(props) => props.theme.blue000};
      }
    `}
`

const BankIcon = styled.div<BankIconProps>`
  height: 30px;
  width: 30px;
  background: url('${(p) => p.url}') 0 0 no-repeat;
  background-size: 30px;
  background-position: center;
`

const BankSearchWrapper = styled.div`
  position: relative;
`
const BankSearchInput = styled.input`
  border: 1px solid ${(p) => p.theme.grey000};
  font-size: 16px;
  width: 100%;
  border-width: 1px 0;
  padding: 20px 0 20px 40px;

  &:active,
  &:focus {
    outline: none;
  }
`

const BankSearchIcon = () => (
  <Icon
    size='20px'
    color='grey600'
    name='magnifier'
    style={{ position: 'absolute', right: '60px', top: '20px' }}
  />
)

const SimpleBankRow = (props: { institution: OBInstitution; onClick: () => void }) => {
  return (
    <BankRow onClick={props.onClick}>
      <div>
        <BankIcon url={props.institution.media[0].source} />
        <Text color='grey900' style={{ marginLeft: '20px' }} weight={600}>
          {props.institution.name}
        </Text>
      </div>
      <Icon
        cursor
        name='chevron-right'
        size='20px'
        color='grey600'
        role='button'
        style={{ marginRight: '24px' }}
      />
    </BankRow>
  )
}

const ModalNavWithBackArrow = (props) => {
  return (
    <NavText color='grey800' size='20px' weight={600}>
      <Icon
        cursor
        name='arrow-back'
        size='20px'
        color='grey600'
        role='button'
        style={{ marginRight: '24px' }}
        onClick={() => props.handleClose()}
      />
      {props.children}
    </NavText>
  )
}

const ModalNavWithCloseIcon = (props) => {
  return (
    <NavText color='grey800' size='20px' weight={600} style={{ justifyContent: 'space-between' }}>
      {props.children}
      <Icon
        cursor
        name='close'
        size='20px'
        color='grey600'
        role='button'
        style={{ marginRight: '24px' }}
        onClick={() => props.handleClose()}
      />
    </NavText>
  )
}

const HrEl = styled.hr`
  border: none;
  border-top: 1px solid ${(p) => p.theme.grey100};
  text-align: center;
  overflow: visible;
  color: #333;
  height: 5px;
  width: 100%;

  &:after {
    content: 'OR';
    padding: 0 4px;
    position: relative;
    top: -10px;
    background: ${(p) => p.theme.alwaysWhite};
  }
`

const Hr = () => {
  return (
    <div style={{ width: '100%' }}>
      <Text weight={600} size='16px' color='grey900'>
        <HrEl />
      </Text>
    </div>
  )
}

const TopText = styled(Text)<{
  marginBottom?: boolean
  spaceBetween: boolean
}>`
  display: flex;
  align-items: center;
  justify-content: ${(props) => (props.spaceBetween ? 'space-between' : 'initial')};
  margin-bottom: ${(props) => (props.marginBottom ? '16px' : '0px')};
`

export {
  BankSearchIcon,
  BankSearchInput,
  BankSearchWrapper,
  BankWaitIndicator,
  BankWrapper,
  BROKERAGE_INELIGIBLE,
  Hr,
  IneligibleErrorMessage,
  LinkOptionsWrapper,
  LinkViaDesktop,
  Loading,
  LoadingUpdating,
  ModalNavWithBackArrow,
  ModalNavWithCloseIcon,
  NavText,
  ScanWithPhone,
  Section,
  SimpleBankRow,
  TopText
}
