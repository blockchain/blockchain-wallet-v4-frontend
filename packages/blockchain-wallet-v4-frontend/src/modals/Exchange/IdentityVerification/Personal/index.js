import React from 'react'
import PropTypes from 'prop-types'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { find, map, propEq } from 'ramda'

import { actions, model } from 'data'
import { getData } from './selectors'
import { Remote } from 'blockchain-wallet-v4'

import Personal from './template'
import Loading from './template.loading'
import DataError from 'components/DataError'

const getCountryElements = countries => [
  {
    group: '',
    items: map(
      country => ({
        value: country,
        text: country.name
      }),
      countries
    )
  }
]

const { PERSONAL_FORM, EMAIL_STEPS } = model.components.identityVerification

class PersonalContainer extends React.PureComponent {
  state = {
    initialEmailVerified: this.props.emailVerified
  }

  componentDidMount () {
    this.fetchData()
  }

  fetchData = () => {
    this.props.actions.fetchSupportedCountries()
    this.props.actions.fetchStates()
  }

  selectAddress = (e, address) => {
    e.preventDefault()
    this.props.actions.selectAddress(address)
  }

  onCountryChange = (e, value) => {
    e.preventDefault()
    this.props.formActions.change(PERSONAL_FORM, 'country', value)
    this.props.formActions.clearFields(PERSONAL_FORM, false, false, 'state')
  }

  editEmail = () => {
    this.props.actions.setEmailStep(EMAIL_STEPS.edit)
  }

  renderForm = ({
    actions,
    initialCountryCode,
    initialEmail,
    email,
    emailVerified,
    emailStep,
    countryCode,
    countryAndStateSelected,
    stateSupported,
    postCode,
    activeField,
    activeFieldError,
    addressRefetchVisible,
    user,
    supportedCountries,
    states,
    handleSubmit
  }) => (
    <Personal
      initialValues={{
        ...user,
        state:
          user.country === 'US'
            ? find(propEq('code', user.state), states) || {}
            : user.state,
        country:
          find(propEq('code', user.country), supportedCountries) ||
          find(propEq('code', initialCountryCode), supportedCountries),
        email: initialEmail
      }}
      showEmail={!this.state.initialEmailVerified}
      emailVerified={emailVerified}
      email={email}
      emailStep={emailStep}
      countryCode={countryCode}
      countryIsUS={countryCode && countryCode === 'US'}
      showStateError={countryAndStateSelected && !stateSupported}
      showPersonal={countryAndStateSelected && stateSupported}
      postCode={postCode}
      supportedCountries={getCountryElements(supportedCountries)}
      states={getCountryElements(states)}
      addressRefetchVisible={addressRefetchVisible}
      activeField={activeField}
      activeFieldError={activeFieldError}
      editEmail={this.editEmail}
      updateEmail={actions.updateEmail}
      sendEmailVerification={actions.sendEmailVerification}
      onAddressSelect={this.selectAddress}
      onCountrySelect={this.onCountryChange}
      onSubmit={handleSubmit}
    />
  )

  render () {
    const { countryData, userData, ...rest } = this.props
    return countryData.cata({
      Success: ({ supportedCountries, states }) =>
        userData.cata({
          Success: user =>
            this.renderForm({ ...rest, supportedCountries, states, user }),
          NotAsked: () => <Loading />,
          Loading: () => <Loading />,
          Failure: () =>
            this.renderForm({ ...rest, supportedCountries, states, user: {} })
        }),
      NotAsked: () => <Loading />,
      Loading: () => <Loading />,
      Failure: () => <DataError onClick={this.props.actions.fetchData} />
    })
  }
}

PersonalContainer.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  initialCountryCode: PropTypes.string,
  supportedCountries: PropTypes.instanceOf(Remote).isRequired,
  countryCode: PropTypes.string
}

PersonalContainer.defaultProps = {
  initialCountryCode: '',
  countryCode: ''
}

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(
    { ...actions.components.identityVerification, ...actions.modules.profile },
    dispatch
  ),
  formActions: bindActionCreators(actions.form, dispatch)
})

export default connect(
  getData,
  mapDispatchToProps
)(PersonalContainer)
