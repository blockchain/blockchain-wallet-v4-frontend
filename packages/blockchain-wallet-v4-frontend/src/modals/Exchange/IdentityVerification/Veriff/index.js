import React from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import styled from 'styled-components'

import { getData } from './selectors'
import { actions } from 'data'
import Loading from './template.loading'
import DataError from 'components/DataError'

const VeriffIframe = styled.iframe.attrs({
  allow: 'camera; microphone'
})`
  width: 100%;
  height: 100%;
  border: none;
`

class Veriff extends React.PureComponent {
  state = {
    loading: false
  }
  componentDidMount () {
    this.props.actions.fetchVeriffUrl()
    window.addEventListener('message', this.handleVeriffMessage, false)
  }

  componentWillUnmount () {
    this.setState({ loading: false })
    window.removeEventListener('message', this.handleVeriffMessage)
  }

  handleVeriffMessage = ({ data, origin }) => {
    const { veriffDomain, actions } = this.props
    if (origin !== veriffDomain) return
    if (data.status !== 'finished') return
    this.setState({ loading: true })
    actions.syncVeriff()
  }

  render () {
    const { veriffUrl, actions } = this.props

    if (this.state.loading) return <Loading />

    return veriffUrl.cata({
      Success: url => (
        <VeriffIframe data-e2e='veriffIframe' src={url} id='veriff-iframe' />
      ),
      Loading: () => <Loading />,
      Failure: () => (
        <DataError data-e2e='veriffFailure' onClick={actions.fetchVeriffUrl} />
      ),
      NotAsked: () => <Loading />
    })
  }
}

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(actions.components.veriff, dispatch),
  kycActions: bindActionCreators(
    actions.components.identityVerification,
    dispatch
  )
})

export default connect(
  getData,
  mapDispatchToProps
)(Veriff)
