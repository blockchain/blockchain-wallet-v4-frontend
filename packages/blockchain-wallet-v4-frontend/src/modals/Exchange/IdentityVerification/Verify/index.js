import React from 'react'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'

import { actions, model } from 'data'
import { getData } from './selectors'
import { MediaContextConsumer } from 'providers/MatchMediaProvider'

import LowFlow from './template.lowflow'
import HighFlow from './template.highflow'
import Loading from './template.loading'
import { hasWebcam } from 'utils/helpers'

const { FLOW_TYPES, KYC_PROVIDERS } = model.components.identityVerification

class VerifyContainer extends React.PureComponent {
  state = {
    showVeriff: false
  }

  componentDidMount () {
    const { actions } = this.props
    actions.fetchSupportedDocuments()
    actions.checkKycFlow()
  }

  showKycProvider = kycProvider => {
    switch (kycProvider) {
      case KYC_PROVIDERS.VERIFF:
        this.setState({ showVeriff: true })
        break
      case KYC_PROVIDERS.ONFIDO:
        this.props.modalActions.showModal(
          'Onfido',
          {
            position: this.props.position + 1,
            total: this.props.total + 1
          },
          {}
        )
        break
      default:
        this.setState({ showVeriff: true })
    }
  }

  render () {
    const { data, actions, modalActions, ...rest } = this.props

    return data.cata({
      Success: ({
        deeplink,
        docTypes,
        email,
        flowConfig,
        needsDocResubmit
      }) => {
        const { flowType, kycProvider } = flowConfig
        return (
          <MediaContextConsumer>
            {({ mobile }) =>
              (flowType === FLOW_TYPES.HIGH && mobile) || !hasWebcam ? (
                <HighFlow
                  email={email}
                  deeplink={deeplink}
                  send={actions.sendDeeplink}
                  done={modalActions.closeAllModals}
                  {...rest}
                />
              ) : (
                <LowFlow
                  supportedDocuments={docTypes}
                  showVeriff={this.state.showVeriff}
                  handleSubmit={() => this.showKycProvider(kycProvider)}
                  needsDocResubmit={needsDocResubmit}
                  {...rest}
                />
              )
            }
          </MediaContextConsumer>
        )
      },
      Loading: () => <Loading />,
      NotAsked: () => null,
      Failure: () => null
    })
  }
}

const mapStateToProps = state => ({
  data: getData(state)
})

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(
    actions.components.identityVerification,
    dispatch
  ),
  modalActions: bindActionCreators(actions.modals, dispatch)
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(VerifyContainer)
