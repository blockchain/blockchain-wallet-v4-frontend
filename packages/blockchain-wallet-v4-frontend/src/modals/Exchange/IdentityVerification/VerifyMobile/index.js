import React from 'react'
import PropTypes from 'prop-types'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { keys } from 'ramda'

import { actions, model } from 'data'
import { getData } from './selectors'

import VerifyMobile from './template'

const { SMS_STEPS } = model.components.identityVerification

class VerifyMobileContainer extends React.PureComponent {
  componentDidMount () {
    const { actions } = this.props
    actions.updateSmsStep()
  }

  render () {
    const {
      activeField,
      activeFieldError,
      smsNumber,
      step,
      countryCode,
      actions,
      handleSubmit,
      onBack
    } = this.props

    return (
      <VerifyMobile
        initialValues={{ smsNumber }}
        smsNumber={smsNumber}
        activeField={activeField}
        activeFieldError={activeFieldError}
        step={step}
        countryCode={countryCode}
        editSmsNumber={actions.setSmsStep.bind(null, SMS_STEPS.edit)}
        updateSmsNumber={actions.updateSmsNumber}
        resendCode={actions.resendSmsCode}
        onSubmit={handleSubmit}
        onBack={onBack}
      />
    )
  }
}

VerifyMobileContainer.propTypes = {
  smsNumber: PropTypes.string.isRequired,
  step: PropTypes.oneOf(keys(SMS_STEPS)).isRequired,
  mobileVerifiedError: PropTypes.string,
  countryCode: PropTypes.object.isRequired,
  handleSubmit: PropTypes.func.isRequired
}

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(actions.components.identityVerification, dispatch)
})

export default connect(
  getData,
  mapDispatchToProps
)(VerifyMobileContainer)
