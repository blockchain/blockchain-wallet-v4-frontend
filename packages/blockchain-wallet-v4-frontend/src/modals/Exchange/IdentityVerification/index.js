import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { compose, bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { contains, keys, pickBy } from 'ramda'
import { FormattedMessage } from 'react-intl'

import { actions, model } from 'data'
import { getData } from './selectors'
import modalEnhancer from 'providers/ModalEnhancer'
import media from 'services/ResponsiveService'
import { ModalHeader, ModalBody } from 'blockchain-info-components'
import Tray, { duration } from 'components/Tray'
import StepIndicator from 'components/StepIndicator'
import DataError from 'components/DataError'
import Loading from './template.loading'
import Personal from './Personal'
import VerifyMobile from './VerifyMobile'
import Verify from './Verify'
import MoreInfo from './MoreInfo'

const HeaderWrapper = styled.div`
  display: flex;
  flex-direction: row;
  ${media.mobile`
    flex-direction: column;
  `};
`

const StepHeader = styled(ModalHeader)`
  padding: 12px !important;
  > div {
    width: 100%;
    > div {
      width: 100%;
    }
  }
  & > :first-child {
    margin-right: 42px;
  }
`
const IdentityVerificationTray = styled(Tray)`
  margin-top: 0;
  border-radius: 0;
  > div:first-child {
    padding: 20px;
    > span:last-child {
      top: 0;
      right: 0;
      margin: 20px;
    }
  }
  > div:last-child {
    overflow: hidden;
    padding: 0;
    height: calc(100% - 57px);
  }
`

const KycStepIndicator = styled(StepIndicator)`
  justify-content: space-between;
  span {
    display: none;
  }
  > img {
    margin-left: 0;
    margin-right: 10px;
    height: 32px;
  }
  > div {
    flex: 1;
    height: 8px;
    max-width: 840px;
    margin: auto;
    padding: 0;
    border-radius: 4px;
    border: none;
    background-color: ${props => props.theme['gray-2']};
    &:after {
      bottom: 0px;
      border-radius: 4px;
      background-color: ${props => props.theme['brand-secondary']};
    }
  }
`

const { STEPS, KYC_MODAL } = model.components.identityVerification

const stepMap = {
  [STEPS.personal]: (
    <FormattedMessage
      id='modals.identityverification.steps.personal'
      defaultMessage='Personal'
    />
  ),
  [STEPS.moreInfo]: (
    <FormattedMessage
      id='modals.identityverification.steps.more_info'
      defaultMessage='Info'
    />
  ),
  [STEPS.mobile]: (
    <FormattedMessage
      id='modals.identityverification.steps.mobile'
      defaultMessage='Phone'
    />
  ),
  [STEPS.verify]: (
    <FormattedMessage
      id='modals.identityverification.steps.verify'
      defaultMessage='Verify'
    />
  )
}

class IdentityVerification extends React.PureComponent {
  state = { show: false }

  componentDidMount () {
    /* eslint-disable */
    this.setState({ show: true })
    /* eslint-enable */
    this.initializeVerification()
  }

  getSteps = steps => pickBy((_, step) => contains(step, steps), stepMap)

  handleClose = () => {
    this.setState({ show: false })
    setTimeout(this.props.close, duration)
  }

  initializeVerification = () => {
    const { tier, isCoinify, needMoreInfo } = this.props
    this.props.actions.initializeVerification(tier, isCoinify, needMoreInfo)
  }

  getStepComponent = step => {
    const { actions } = this.props
    if (step === STEPS.personal)
      return (
        <Personal
          handleSubmit={actions.savePersonalData}
          onBack={actions.goToPrevStep}
        />
      )

    if (step === STEPS.moreInfo) return <MoreInfo />

    if (step === STEPS.mobile)
      return (
        <VerifyMobile
          handleSubmit={actions.verifySmsNumber}
          onBack={actions.goToPrevStep}
        />
      )

    if (step === STEPS.verify) return <Verify onBack={actions.goToPrevStep} />
  }

  render () {
    const { show } = this.state
    const { step, steps, position, total } = this.props

    return (
      <IdentityVerificationTray
        in={show}
        class='tray'
        position={position}
        total={total}
        onClose={this.handleClose}
        data-e2e='identityVerificationModal'
      >
        {steps.cata({
          Success: steps => (
            <React.Fragment>
              <StepHeader
                tray
                paddingHorizontal='15%'
                onClose={this.handleClose}
              >
                <HeaderWrapper>
                  <KycStepIndicator
                    adjuster={0.1667}
                    barFullWidth
                    horizontalMobile
                    flexEnd
                    maxWidth='none'
                    step={step}
                    stepMap={this.getSteps(steps)}
                  />
                </HeaderWrapper>
              </StepHeader>
              <ModalBody>{this.getStepComponent(step)}</ModalBody>
            </React.Fragment>
          ),
          Loading: () => <Loading />,
          NotAsked: () => <Loading />,
          Failure: () => <DataError onClick={this.initializeVerification} />
        })}
      </IdentityVerificationTray>
    )
  }
}

IdentityVerification.propTypes = {
  step: PropTypes.oneOf(keys(STEPS)),
  close: PropTypes.func.isRequired
}

IdentityVerification.defaultProps = {
  step: STEPS.personal
}

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(actions.components.identityVerification, dispatch)
})

const enhance = compose(
  modalEnhancer(KYC_MODAL, { preventEscapeClose: true }),
  connect(
    getData,
    mapDispatchToProps
  )
)

export default enhance(IdentityVerification)
