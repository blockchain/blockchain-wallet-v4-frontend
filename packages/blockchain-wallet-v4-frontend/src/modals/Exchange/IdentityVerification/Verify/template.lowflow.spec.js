import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { values } from 'ramda'

import { model } from 'data'
import LowFlow from './template.lowflow'
import { BackButton } from 'components/IdentityVerification'

const handleSubmit = jest.fn()
const onBack = jest.fn()

const { SUPPORTED_DOCUMENTS } = model.components.identityVerification

const supportedDocuments = values(SUPPORTED_DOCUMENTS)

describe('HighFlow', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('should render correctly', () => {
    const component = shallow(
      <LowFlow
        supportedDocuments={supportedDocuments}
        onBack={onBack}
        handleSubmit={handleSubmit}
      />
    )
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })

  it('should trigger onBack on back button click', () => {
    const component = shallow(
      <LowFlow
        supportedDocuments={supportedDocuments}
        onBack={onBack}
        handleSubmit={handleSubmit}
      />
    )
    component
      .find('FooterShadowWrapper')
      .dive()
      .find(BackButton)
      .simulate('click')
    expect(onBack).toHaveBeenCalledTimes(1)
  })

  it('should trigger submit on submit click', () => {
    const component = shallow(
      <LowFlow
        supportedDocuments={supportedDocuments}
        onBack={onBack}
        handleSubmit={handleSubmit}
      />
    )
    component
      .find('FooterShadowWrapper')
      .dive()
      .find('Button[nature="primary"]')
      .simulate('click')
    expect(handleSubmit).toHaveBeenCalledTimes(1)
  })
})
