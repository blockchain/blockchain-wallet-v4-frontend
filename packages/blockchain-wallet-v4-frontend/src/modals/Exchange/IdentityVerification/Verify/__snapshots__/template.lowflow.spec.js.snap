// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HighFlow should render correctly 1`] = `
<IdentityVerification__IdentityVerificationForm>
  <FooterShadowWrapper
    fields={
      <ForwardRef>
        <ForwardRef>
          <ForwardRef>
            <FormattedMessage
              defaultMessage="Last Step. Let's Confirm It's You"
              id="identityverification.verify.header.confirm"
              values={Object {}}
            />
            <ForwardRef>
              <TooltipHost
                id="identityverification.headerhelper"
              >
                <ForwardRef
                  color="brand-primary"
                  name="question-in-circle-filled"
                  size="24px"
                />
              </TooltipHost>
              <ForwardRef
                name="identity-verification"
              />
            </ForwardRef>
          </ForwardRef>
          <ForwardRef>
            <FormattedMessage
              defaultMessage="We need to confirm your identity with a government issued ID and selfie. Before proceeding, make sure you have one of the following forms of ID handy and your camera is enabled."
              id="identityverification.verify.messagewithcamera"
              values={Object {}}
            />
          </ForwardRef>
          <ForwardRef>
            <FormattedMessage
              defaultMessage="Government Issued Passport"
              id="identityverification.verify.passport"
              values={Object {}}
            />
            <FormattedMessage
              defaultMessage="Driverâ€™s License"
              id="identityverification.verify.driverslicense"
              values={Object {}}
            />
            <FormattedMessage
              defaultMessage="National Identity Card"
              id="identityverification.verify.natitionalidcard"
              values={Object {}}
            />
          </ForwardRef>
          <ForwardRef>
            <Text
              altFont={false}
              capitalize={false}
              color="gray-5"
              cursor="inherit"
              italic={false}
              opacity={1}
              size="18px"
              uppercase={false}
              weight={400}
            >
              <FormattedMessage
                defaultMessage="How do I enable my camera and microphone?"
                id="identityverification.verify.how_to_camera_mic"
                values={Object {}}
              />
            </Text>
            <Text
              altFont={false}
              capitalize={false}
              color="gray-5"
              cursor="inherit"
              italic={false}
              opacity={1}
              size="16px"
              uppercase={false}
              weight={300}
            >
              <FormattedMessage
                defaultMessage="Click allow when prompted above or enable in your browser settings."
                id="identityverification.verify.instructions"
                values={Object {}}
              />
            </Text>
            <Image
              color="auto"
              height="auto"
              name="allow-camera"
              srcset={
                Object {
                  "allow-camera2": "2x",
                  "allow-camera3": "3x",
                }
              }
              width="auto"
            />
          </ForwardRef>
        </ForwardRef>
      </ForwardRef>
    }
    footer={
      <ForwardRef>
        <ForwardRef
          data-e2e="lowflowBackButton"
          onClick={[MockFunction]}
        >
          <FormattedMessage
            defaultMessage="Back"
            id="identityverification.lowflow.personal.back"
            values={Object {}}
          />
        </ForwardRef>
        <Button
          bold={false}
          capitalize={false}
          data-e2e="lowflowContinueButton"
          disabled={false}
          fullwidth={false}
          height="40px"
          nature="primary"
          onClick={[MockFunction]}
          rounded={false}
          size="14px"
          small={false}
          uppercase={false}
        >
          <FormattedMessage
            defaultMessage="Continue"
            id="identityverification.lowflow.personal.continue"
            values={Object {}}
          />
        </Button>
      </ForwardRef>
    }
  />
</IdentityVerification__IdentityVerificationForm>
`;
